connectivity/check/policy.go

```diff
@@ -0,0 +1,185 @@
+// be reported like any other test results.
+type CiliumNetworkPolicy struct {
+	Title  string
+	Policy string
+}
+
+// Name returns the absolute name of the policy
+func (p *CiliumNetworkPolicy) Name() string {
+	return p.Title
+}
+
+// Run applies the policy, use nil policy to delete all policies
+func (p *CiliumNetworkPolicy) Run(ctx context.Context, c TestContext) {
+	failures := c.ApplyPolicyYaml(ctx, p.Policy)
+	c.Report(TestResult{
+		Name:     p.Name(),
+		Failures: failures,
+		Warnings: 0,
+	})
+}
+
+type WithPolicy struct {
+	err  error
+	cnps []*ciliumv2.CiliumNetworkPolicy
+}
+
+func (wp *WithPolicy) Parse(policy string) {
+	wp.cnps, wp.err = ParsePolicyYaml(policy)
+}
+
+// Apply returns the number of failures and a cancel function that deletes the applied policies
+func (wp *WithPolicy) Apply(ctx context.Context, c TestContext) (failures int, cancel func()) {
+	if wp.err != nil {
+		return 1, func() {}
+	}
+
+	for _, cnp := range wp.cnps {
+		failures += c.ApplyCNP(ctx, cnp)
+	}
+	return failures, func() {
+		for _, cnp := range wp.cnps {
+			c.DeleteCNP(ctx, cnp)
+		}
+	}
+}
+
+// Run applies the policy, use empty policy to delete all policies
+func (k *K8sConnectivityCheck) ApplyPolicyYaml(ctx context.Context, policy string) int {
+	if policy == "" {
+		// Delete all policies
+		return k.ApplyCNP(ctx, nil)
+	}
+	failures := 0
+	cnps, err := ParsePolicyYaml(policy)
+	if err != nil {
+		k.Log("❌ %s", err)
+		failures++
+	} else {
+		for _, cnp := range cnps {
+			failures += k.ApplyCNP(ctx, cnp)
+		}
+	}
+	return failures
+}
+
+// ParsePolicyYaml decodes policy yaml into a slice of CiliumNetworkPolicies
+func ParsePolicyYaml(policy string) (cnps []*ciliumv2.CiliumNetworkPolicy, err error) {
+	if policy == "" {
+		return nil, nil
+	}
+	yamls := strings.Split(policy, "---")
+	for _, yaml := range yamls {
+		if yaml == "\n" || yaml == "" {
+			continue
+		}
+		obj, groupVersionKind, err := scheme.Codecs.UniversalDeserializer().Decode([]byte(yaml), nil, nil)
+		if err != nil {
+			return nil, fmt.Errorf("Resource decode error (%s) in: %s", err, yaml)
+		}
+		switch groupVersionKind.Kind {
+		case "CiliumNetworkPolicy":
+			cnp, ok := obj.(*ciliumv2.CiliumNetworkPolicy)
+			if !ok {
+				return nil, fmt.Errorf("Object cast to CiliumNetworkPolicy failed: %s", yaml)
+			}
+			cnps = append(cnps, cnp)
+		default:
+			return nil, fmt.Errorf("Unknown policy type '%s' in: %s", groupVersionKind.Kind, yaml)
+		}
+	}
+	return cnps, nil
+}
+
+// DeleteCNP deletes a CNP
+func (k *K8sConnectivityCheck) DeleteCNP(ctx context.Context, cnp *ciliumv2.CiliumNetworkPolicy) {
+	name := cnp.Namespace + "/" + cnp.Name
+	if err := k.deleteCNP(ctx, cnp); err != nil {
+		k.Log("❌ [%s] policy delete failed: %s", name, err)
+	}
+	delete(k.policies, name)
+}
+
+// ApplyCNP returns the number of failures
+func (k *K8sConnectivityCheck) ApplyCNP(ctx context.Context, cnp *ciliumv2.CiliumNetworkPolicy) int {
```

Comment:
`ApplyCNP()` is called from loops where we keep failure counts. These call sites are simpler if this returns a failure count as well. Since there is no loop here I changed this to return 0 or 1 directly as you suggested above :-)